File: A01 Array/Search roated array/Leetcode33 Search in Rotated Sorted Array/Leetcode 33.java
/*
O(long n)
Space: O(1)

*/ 


class Solution {
    public int search(int[] nums, int target) {
        if(nums == null || nums.length == 0) return -1; 
        int peak_index = findPeakIndex(nums); 
        if (peak_index >= 0 && target >= nums[0] && target <= nums[peak_index]){  //
            return binarySearch(nums, 0, peak_index, target);
        }else{
            return binarySearch(nums, peak_index +1, nums.length -1, target);
        }
    }
    
    private int findPeakIndex(int[] nums){
        if (nums.length == 1) return 0;                 //
        if (nums[0] < nums[nums.length -1]) return nums.length -1;    //Ascending order
        int left = 0, right = nums.length -1; 
        while(left <= right){
            int mid = left + (right - left) /2;
            if(nums[mid] > nums[mid +1]) return mid; 
            else if (nums[left] <= nums[mid]){
                left = mid +1; 
            } else{
                right = mid -1;
            }
        }
        return 0; 
    }
    
    private int binarySearch(int[] nums, int left, int right, int target){
        while(left <= right){
            int mid = left + (right - left) /2;
            if(nums[mid]  == target) return mid; 
            else if (nums[mid] >= target){
                right = mid -1; 
            } else{
                left = mid +1;
            }
        }
        return -1; 
    }
}

File: A01 Array/Search roated array/Leetcode33 Search in Rotated Sorted Array/Leetcode 33.py
#############
Binary search to find index of pivot (where the array has been rotated).
Normal Binary search on both sides of pivot to find element
Code uses the property that the last element of array will be smaller than all elements before the pivot. See example below to clarify:
example: in rotated array: 8,9,11,5,6,7. We know 5 is the pivot.
All elements on left side of pivot are larger than last element, 7
All elements on right side of pivot including pivot are smaller than last element, 7.
We can use this property to binary search and find pivot, once we have pivot, run normal bin_search on both sides of pivot and we're done woohoo!
##############


class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if(not nums):
            return -1
        pivot_index = self.find_pivot(nums) 
        result = self.bin_search(nums, target, 0, pivot_index-1)
        if(result!=-1):
            return result
        else:
            return self.bin_search(nums, target, pivot_index, len(nums)-1)
        
    def find_pivot(self, arr):
	#compare last element to mid element. 
	#If mid element is greater than last element, pivot must be on right move low to mid+1
	#If mid element is less than last element, move high to mid-1
        element_to_compare = arr[-1]
        low = 0
        high = len(arr)-1    
        while(low <= high):
            mid = (low+high)//2
            if element_to_compare < arr[mid]:
                low = mid+1
            elif element_to_compare >= arr[mid]:
                high = mid-1
        return low    

    def bin_search(self, arr, value, low, high):     
        while(low <= high):
            mid = (low+high)/2
            if value<arr[mid]:
                high = mid-1
            elif value>arr[mid]:
                low = mid+1
            else:
                return mid
        return -1